Wrappers/Python/cil/optimisation/functions/Function.py

```diff
@@ -320,7 +326,88 @@ def __add__(self, other):
         else:
             return super(SumFunction, self).__add__(other)
 
+class SubsetSumFunction(SumFunction):
+
+    r"""SubsetSumFunction represents the following sum 
+    
+    .. math:: \frac{1}{n}\sum_{i=1}^{n} F_{i} = \frac{1}{n}(F_{1} + F_{2} + ... + F_{n})
+
+    where :math:`n` is the number of subsets.
+
+    Parameters:
+    -----------
+    functions : list(functions) 
+                A list of functions: :code:`[F_{1}, F_{2}, ..., F_{n}]`. Each function is assumed to be smooth function with an implemented :func:`~Function.gradient` method.
+    sampling : :obj:`string`, Default = :code:`random`
+               Selection process for each function in the list. It can be :code:`random` or :code:`sequential`. 
+    replacement : :obj:`boolean`. Default = :code:`True`
+               The same subset can be selected when :code:`replacement=True`. 
+                
+    Example
+    -------
+
+    .. math:: \sum_{i=1}^{n} F_{i}(x) = \sum_{i=1}^{n}\|A_{i} x - b_{i}\|^{2}
+
+    >>> f = SubsetSumFunction([LeastSquares(Ai, b=bi)]*n for Ai,bi in zip(A_subsets, b_subsets))   
+  
+    """
+    
+    def __init__(self, functions, 
+                 sampling = "random",
+                 replacement = True,
+                 subset_init = -1):    
+        
+        super(SubsetSumFunction, self).__init__(*functions)
+        
+        self.subset_num = subset_init
+        self.data_passes = [0]
+        self.sampling = sampling  
+        self.replacement = replacement
+        self.enumerate_functions = list(range(self.num_subsets))
+                                                            
+    @property
+    def num_subsets(self):
+        return self.num_functions
 
+    def __call__(self, x):
+        r"""Returns the value of the averaged sum of functions at :math:`x`.		
+        		
+        .. math:: \frac{1}{n}(F_{1} + F_{2} + ... + F_{n}))(x) = \frac{1}{n} * ( F_{1}(x) + F_{2}(x) + ... + F_{n}(x))
+        """
+                		
+        return (1/self.num_subsets) * super(SubsetSumFunction, self).__call__(x)      
+        
+    def _full_gradient(self, x, out=None):
+        r""" Computes the full gradient at :code:`x`. It is the sum of all the gradients for each function. """
```

Comment:
doc is correct now. still remains why we'd call this `_full_gradient`

Reply:
`_full_gradient` is used in `SVRG` although I am not sure if it is needed, see [here](https://github.com/TomographicImaging/Hackathon-000-Stochastic-Algorithms/blob/286c37eaa09fdf76c3c4ca7585f645277edbbf81/cil/NewSubsetSumFunction.py#L573), @zeljkozeljko any ideas?

Reply:
the "old" `SubsetSumFunction` class had a `_full_gradient` method, and SVRG had an attribute called `full_gradient`, which was computed (updated) once every couple of epochs and it is the estimator of the full gradient of the objective at the snapshot/anchored image. This attribute is then used at every update and I would think we want to store (or at least have the option of storing it), and not recompute it at every update (otherwise why not use full GD). As to the naming, i'm not sure if the choice of the underscore in the method  was because of some CIL convention, but of course the, as regards to the attribute in SVRG, this can be renamed (eg we could have `snapshot_full_gradient` and then also `snapshot_image`, or something similar), and reserve `full_gradient` for the `SubsetSumFunction` class

Reply:
can be resolved now